### ISSUE TITLE
Environment Variables Referenced in supervisor.conf

### ISSUE THREAD

**Issue by jcrawford** [Supervisor/supervisor#1380]
I had found this closed issue, and it is exactly what I am experiencing however, it states the fix was in 3.2.0 and that is the version I am running, yet I cannot get the reference to work properly.

https://github.com/Supervisor/supervisor/issues/750

I am currently running Supervisor 3.2.0 and I cannot seem to get this to work properly. I have tried several methods and did try the suggested solution in the referenced issue above.

```
[program:laravel-worker]
process_name=%(program_name)s_%(process_num)02d
command=/usr/bin/php /var/www/html/artisan queue:work
autostart=true
autorestart=true
stdout_logfile=/var/log/worker.log
numprocs=%(ENV_WORKER_THREADS)s
```
in my `.env` file I have `WORKER_THREADS=2`. I even checked that it is in the shell

```
$ echo $WORKER_THREADS
2
```
When I try to run `supervisorctl reread` I get the following error.

```
ERROR: CANT_REREAD: Format string '%(ENV_WORKER_THREADS)s' for 'program:laravel-worker.numprocs' contains names ('ENV_WORKER_THREADS') which cannot be expanded. Available names: ENV_HOME, ENV_HOSTNAME, ENV_OLDPWD, ENV_PATH, ENV_PWD, ENV_SHLVL, ENV_TERM, ENV__, group_name, here, host_node_name, program_name in section 'program:laravel-worker' (file: '/etc/supervisor/conf.d/supervisor.conf')
```

I have tried using `%(WORKER_THREADS)s` and `%(ENV_WORKER_THREADS)s` and neither seems to work for me.

Any thoughts?

**Comment by mnaberez** [comment]
Supervisor 3.2.0 was released in [2015](https://github.com/Supervisor/supervisor/blob/3.2.0/CHANGES.txt#L1) and is no longer maintained.  Here is what happens on 4.2.1:

```
$ cat supervisord.conf 
[supervisord]

[program:cat]
command = /bin/cat 
process_name=%(program_name)s_%(process_num)02d
numprocs=%(ENV_WORKER_THREADS)s

$ supervisord --version
4.2.1

$ supervisord -n -c supervisord.conf 
Error: Format string '%(ENV_WORKER_THREADS)s' for 'program:cat.numprocs' contains names 
('ENV_WORKER_THREADS') which cannot be expanded. Available names: ENV_HOME, ..., 
group_name, here, host_node_name, program_name in section 'program:cat' (file: 'supervisord.conf')

$ export WORKER_THREADS=3

$ supervisord -n -c supervisord.conf 
2020-10-20 10:57:25,595 INFO Increased RLIMIT_NOFILE limit to 1024
2020-10-20 10:57:25,600 INFO supervisord started with pid 19117
2020-10-20 10:57:26,607 INFO spawned: 'cat_00' with pid 19119
2020-10-20 10:57:26,610 INFO spawned: 'cat_01' with pid 19120
2020-10-20 10:57:26,613 INFO spawned: 'cat_02' with pid 19121
```

When the environment variable exists in the environment at the time `supervisord` is started, then it will be available to be used in expansions in the config file.  

Above first shows that the `WORKER_THREADS` variable did not exist in the environment when `supervisord` was started, and it produced a similar error message to the original report.  It then shows that when the `WORKER_THREADS` variable exists in the environment when `supervisord` is started, it is available in the expansions.

Please note that `supervisord` does not have any special integration with any particular operating system, so it does not know about special files like `/etc/environment`, `.env`, and does not start a login shell.  A more complete description can be found in [Subprocess Environment](http://supervisord.org/subprocess.html#subprocess-environment).

**Comment by jcrawford** [comment]
I am not quite sure why this was closed because the issue was not addressed, however, it did turn out to be an issue on my side.  We were loading our environment file from laravel with `. .env` however I noticed that this was not putting the environment variables into the `env` command output nor the output of `printenv` and after reviewing the source code from the PR mentioned in #750 I can see that is where it is trying to get the values from.

I now need to figure out how to get our `.env` file into the actual system environment and not just the shell environment because supervisor is running under it's own process and will not have access to these environment variables.

**Comment by jcrawford** [comment]
I have actually been able to do `export $(cat .env | xargs)` and get all environment variables to appear in `env` and `printenv` however, even after doing `supervisorctl restart all` it does not seem to want to expand my variables, I am just trying to understand what I might be doing wrong.

I simply have the following

```
WORKER_THREADS=2
```

in my supervisor.conf I have 

```
numprocs=%(ENV_WORKER_THREADS)s
```

when I run `env` and `printenv` I see the following

```
WORKER_THREADS=2
```

when I do `echo $WORKER_THREADS`

```
$ echo $WORKER_THREADS
2
```

When I restart supervisor and then try to reread the configuration

```
$ supervisorctl restart all
nginx: stopped
laravel-worker:laravel-worker_00: stopped
nginx: started
laravel-worker:laravel-worker_00: started
```

```
$ supervisorctl reread
ERROR: CANT_REREAD: Format string '%(ENV_WORKER_THREADS)s' for 'program:laravel-worker.numprocs' contains names ('ENV_WORKER_THREADS') which cannot be expanded. Available names: ENV_HOME, ENV_HOSTNAME, ENV_OLDPWD, ENV_PATH, ENV_PWD, ENV_SHLVL, ENV_TERM, ENV__, group_name, here, host_node_name, program_name in section 'program:laravel-worker' (file: '/etc/supervisor/conf.d/supervisor.conf')

I have tried this on 3.2.0 and 4.1.0 on ubuntu so your assistance would be greatly appreciated.

**Comment by jcrawford** [comment]
@mnaberez I am not quite sure if you will find this helpful or not since you wanted to simply close the issue without any further investigation however it seems that I have solved the issue.

When I start supervisor I was starting it with `supervisord -n` and this started up fine and was using the configuration file under /etc/supervisor/conf.d/supervisor.conf however, it would not expand the environment variables.

In order to solve this and make it work I had to start supervisor by specifying the configuration file.

`supervisord -n -c /etc/supervisor/conf.d/supervisor.conf`

Doing this it would start up properly and expand the environment variable.  Without specifying the configuration file it was still using the same file (as seen by the errors) but it would not expand the variable.
